northstar/src/runtime/process/mod.rs

```diff
@@ -272,87 +320,50 @@ impl super::state::Process for Process {
     }
 }
 
-/// Spawn a task that waits for the process to exit. Resolves to the exit status of `pid`.
-fn waitpid(
-    container: Container,
-    pid: Pid,
-    tx: EventTx,
-    stop: CancellationToken,
-) -> impl Future<Output = ExitStatus> {
-    let mut sigchld = signal::unix::signal(signal::unix::SignalKind::child())
-        .expect("Failed to set up signal handle for SIGCHLD");
-
-    task::spawn(async move {
-        // Check the status of the process after every SIGCHLD is received
-        let exit_status = loop {
-            sigchld.recv().await;
-            if let Some(exit) = exit_status(pid) {
-                stop.cancel();
-                break exit;
-            }
-        };
-
-        drop(
-            tx.send(Event::Container(
-                container,
-                ContainerEvent::Exit(exit_status.clone()),
-            ))
-            .await,
-        );
-        exit_status
-    })
-    .map(|r| r.expect("Task join error"))
-}
-
-/// Get exit status of process with `pid` or None
-fn exit_status(pid: Pid) -> Option<ExitStatus> {
+/// Get exit status of process with `pid` if the process already exited. Panic otherwise.
+fn init_exit_status(pid: Pid) -> ExitStatus {
     let pid = unistd::Pid::from_raw(pid as i32);
-    match wait::waitpid(Some(pid), Some(WaitPidFlag::WNOHANG)) {
+    let result = wait::waitpid(Some(pid), None);
```

Comment:
With the `WNOHANG` we can have the situation that the init process closed the pipe but it not yet exited.
Yes - this blocks as we block in some other places as well. The assumption is that in `init` the pipe is never closed and the only possibility that the read makes progress is that init sends a exit stati of the child or exits itself. In both cases the `waitpid` is fine.

Reply:
hmm okay but shouldn't we at least use some `spawn_blocking` or `block_in_place`?

Reply:
We need to run on a single threaded runtime because we `clone`. There's no `spawn_blocking` or `block_in_place`.
The code in in init is supposed to work. If it's not - and init crashes - the `waitpid` also returns. The only chance that we hang forever is when the `waitpid` in `init` hangs forever - but then we're screwed anyways....

Reply:
okay, looks like there is no way around then :sweat_smile: